clients/src/test/java/org/apache/kafka/clients/consumer/internals/ListOffsetsRequestManagerTest.java

```diff
@@ -0,0 +1,217 @@
+    private static final Node LEADER_1 = new Node(0, "localhost", 9092);
+    private static final IsolationLevel DEFAULT_ISOLATION_LEVEL = IsolationLevel.READ_COMMITTED;
+
+    @BeforeEach
+    public void setup() {
+        metadata = mock(ConsumerMetadata.class);
+        subscriptionState = mock(SubscriptionState.class);
+        this.time = new MockTime(0);
+        requestManager = new ListOffsetsRequestManager(subscriptionState,
+                metadata, DEFAULT_ISOLATION_LEVEL, mock(Time.class),
+                mock(ApiVersions.class), new LogContext());
+    }
+
+    @Test
+    public void testListOffsetsRequest_Success() throws ExecutionException, InterruptedException {
+        Map<TopicPartition, Long> partitionsOffsets = new HashMap<>();
+        TopicPartition tp = new TopicPartition(TEST_TOPIC, TEST_PARTITION);
+        long offset = ListOffsetsRequest.EARLIEST_TIMESTAMP;
+        partitionsOffsets.put(tp, offset);
+
+        expectSuccessfulRequest();
+        CompletableFuture<Map<TopicPartition, Long>> result = requestManager.fetchOffsets(
+                partitionsOffsets.keySet(),
+                ListOffsetsRequest.EARLIEST_TIMESTAMP,
+                false);
+        assertEquals(1, requestManager.requestsToSend());
+        assertEquals(0, requestManager.requestsToRetry());
+
+        NetworkClientDelegate.PollResult res = requestManager.poll(time.milliseconds());
+        verifySuccessfulPoll(res);
+
+        NetworkClientDelegate.UnsentRequest unsentRequest = res.unsentRequests.get(0);
+        unsentRequest.future().complete(buildListOffsetsClientResponse(unsentRequest,
+                partitionsOffsets));
+
+        verifyRequestSuccessfullyCompleted(tp, offset, result, partitionsOffsets);
+    }
+
+    @Test
+    public void testListOffsetsRequest_FailAndRetrySuccess() throws ExecutionException, InterruptedException {
+        Map<TopicPartition, Long> partitionsOffsets = new HashMap<>();
+        TopicPartition tp = new TopicPartition(TEST_TOPIC, TEST_PARTITION);
+        long offset = 1L;
+        partitionsOffsets.put(tp, offset);
+
+        // List offsets request that fails with unknown leader
+        expectFailedRequest_MissingLeader();
+        CompletableFuture<Map<TopicPartition, Long>> fetchOffsetsFuture =
+                requestManager.fetchOffsets(partitionsOffsets.keySet(),
+                        ListOffsetsRequest.EARLIEST_TIMESTAMP,
+                        false);
+        assertEquals(0, requestManager.requestsToSend());
+        assertEquals(1, requestManager.requestsToRetry());
+        NetworkClientDelegate.PollResult res = requestManager.poll(time.milliseconds());
+        assertEquals(0, res.unsentRequests.size());
+        assertFalse(fetchOffsetsFuture.isDone());
+
+        // Mock metadata update. Previously failed request should be retried and succeed
+        expectSuccessfulRequest();
+        requestManager.onUpdate(new ClusterResource(""));
```

Comment:
is it possible for us to actually trigger metadata.update() to trigger the onUpdate callback? Can we invoke networkClientDelegate.poll to trigger networkClient.poll (using mockClient), which should invoke metadata.update and trigger the requestManager.onUpdate() ? I'm guessing we might use MockClient and invoke networkClientDelegate.poll() to eventually the onUpdate.  WDYT?

Reply:
I guess this can be on the BackgroundThread level testing.

Reply:
exactly, it's definitely something important to test but I wouldn't put it here. The ListOffsetsRequestManager is not responsible for triggering the metadata update or updating the in-memory metadata cache after it. It is responsible of properly resuming its job after the update has been done (or not done), so that's what we are trying to test here. More tests to come still but all based on the metadata update outcome. 

I will definitely include it on the BackgroundThread level testing.

Reply:
Tracked on [KJC-537](https://confluentinc.atlassian.net/browse/KJC-537) for reviewing test coverage of the metadata update/listeners triggering logic on the appropriate component.